### DISCUSSION TITLE
cRLSign key usage enforcement

### DISCUSSION THREAD

**Discussion by edmorris-github** [strongswan/strongswan#1548] [General]
In this old issue found here ( https://wiki.strongswan.org/issues/3529 ) Tobias states that "... accepting a CRL that's issued by a CA without cRLSign keyUsage is in compliance with RFC 5280. The keyUsage is basically to allow non-CA certificates to sign CRLs. The latter is what the openssl plugin currently doesn't allow for some reason (I guess Noel is right that it was simply overlooked)."

I don't believe that this is correct.  RFC 5280 Section 4.2.1.3 (Key Usage) states,

"
   Conforming CAs MUST include this [Key Usage] extension in certificates that
   contain public keys that are used to validate digital signatures on
   other public key certificates or CRLs.
"
"
   If the keyUsage extension is present, then the subject public key
   MUST NOT be used to verify signatures on certificates or CRLs unless
   the corresponding keyCertSign or cRLSign bit is set.
"

My understanding is that a CA that is not also a CRL issuer may (assuming it's not providing revocation information through an OCSP responder) delegate CRL signing to another entity.  And the feature to allow this is the cRLIssuer field of the CRLDistributionPoints extension (and the corresponding indirectCRL flag in the IssuingDistributionPoint of a CRL), and not the cRLSign key usage bit alone.

Thus, CRL path validation should enforce that (in a simple case where indirectCRLs are not used) CAs that are also CRL issuers have the cRLSign key usage bit.  This issue comes up periodically as government certifications check for the presence of the cRLSign bit for CAs that issue CRLs.

Thanks
Ed

**Comment by tobiasbrunner** [comment]
> I don't believe that this is correct.

If you are referring to the first part of that quote, then yes, technically, that's only correct if the CA certificate contained no keyUsage extension (which our code currently can't differentiate at the location of the patch that was referenced in that issue).

> and not the cRLSign key usage bit alone.

I did not say anything like that. I was referring to the specific patch that allowed CRLs, on a very basic level, unrelated to CDPs, to be issued by non-CA certificates that have the cRLSign key usage bit set. Before that patch, the _openssl_ plugin only accepted CA certificates as CRL issuer.

> Thus, CRL path validation should enforce that (in a simple case where indirectCRLs are not used) CAs that are also CRL issuers have the cRLSign key usage bit.

Yeah, I guess, that's true according to section 6.3.3. (f):

      If a key usage extension is present in the CRL issuer's certificate,
      verify that the cRLSign bit is set.

Not sure if there is any benefit in enforcing that for CA certificates, though. An as mentioned, at the point where the CRL or its signature is verified, we currently don't have the information whether the keyUsage extension was present in the issuing certificate. Note that we also don't use the keyCertSign key usage bit to decide whether a CA is allowed to issue certificates, in both cases we currently simply rely on the cA basic constraint and allow all such certificates to issue certificates and CRLs.

> This issue comes up periodically as government certifications check for the presence of the cRLSign bit for CAs that issue CRLs.

What do you mean? Are you referring to something like test cases that use CRLs signed by CA certificates that don't have the cRLSign key usage bit that strongSwan passes when it technically shouldn't?

**Reply by edmorris-github** [reply]
> If you are referring to the first part of that quote, then yes, technically, that's only correct if the CA certificate contained no keyUsage extension (which our code currently can't differentiate at the location of the patch that was referenced in that issue).

A CA certificate (used to issue certs or CRLs) that lacks the keyUsage extension would fail to comply with RFC 5280, agreed?  Understood regarding the code location.  Often, an implementation no longer has the CA chain from the (typically) prior path validation, and an implementation cannot rely upon the presence of an AIA caIssuers reference.

> Not sure if there is any benefit in enforcing that for CA certificates, though. An as mentioned, at the point where the CRL or its signature is verified, we currently don't have the information whether the keyUsage extension was present in the issuing certificate. Note that we also don't use the keyCertSign key usage bit to decide whether a CA is allowed to issue certificates, in both cases we currently simply rely on the cA basic constraint and allow all such certificates to issue certificates and CRLs.

Agreed about the difficulty in referencing the issuing certificate during CRL checking (our test PKI certificates always include id-ad-caIssuers to be "nice" for this very reason).  Using the cA basic constraints simplifies the logic and while perhaps sensible, doesn't conform to the RFC. 

> What do you mean? Are you referring to something like test cases that use CRLs signed by CA certificates that don't have the cRLSign key usage bit that strongSwan passes when it technically shouldn't?

Yes, exactly (our test case and cert chains are described in the original issue details).  Having an option (configuration flag?) to enforce strict RFC 5280 compliance would satisfy the certification testing (which specifies the RFC).

**Reply by tobiasbrunner** [reply]
> A CA certificate (used to issue certs or CRLs) that lacks the keyUsage extension would fail to comply with RFC 5280, agreed?

Sure, but it's still an optional extension. And the code is old, so it's possible that there were PKIs that didn't add it at the time (even if it was already a MUST in RFC 3280, however, it seems it wasn't in RFC 2459). Actually, strongSwan didn't add it either when creating CA certificates before 1ec8f22de2229520b97981224d7655f4d9e5b1e9.

> > What do you mean? Are you referring to something like test cases that use CRLs signed by CA certificates that don't have the cRLSign key usage bit that strongSwan passes when it technically shouldn't?
> 
> Yes, exactly (our test case and cert chains are described in the original issue details). Having an option (configuration flag?) to enforce strict RFC 5280 compliance would satisfy the certification testing (which specifies the RFC).

I guess we could handle this similar to the RFC 4945 compliance and just set the internal `X509_CRL_SIGN` flag if the extension is absent in CA certificates and then always enforce the flag when verifying the signature. I've pushed some commits to that end to the _1548-crl-sign_ branch. The only issue could be actual CA certificates with keyUsage extension and unset cRLSign bit that issue CRLs in the wild.

**Reply by edmorris-github** [reply]
> Sure, but it's still an optional extension. And the code is old, so it's possible that there were PKIs that didn't add it at the time (even if it was already a MUST in RFC 3280, however, it seems it wasn't in RFC 2459). Actually, strongSwan didn't add it either when creating CA certificates before https://github.com/strongswan/strongswan/commit/1ec8f22de2229520b97981224d7655f4d9e5b1e9.

1999 and 2002 were still the early days of X.509, and until ~2017-2019 there were still some v1 Root CAs included in Browsers and OS trust anchor databases.  I believe that most of those legacy CAs have expired as of today.

> I guess we could handle this similar to the RFC 4945 compliance and just set the internal X509_CRL_SIGN flag if the extension is absent in CA certificates and then always enforce the flag when verifying the signature. I've pushed some commits to that end to the 1548-crl-sign branch. The only issue could be actual CA certificates with keyUsage extension and unset cRLSign bit that issue CRLs in the wild.

Such a change sounds reasonable.  Regarding certificates in the wild, would it be safe to put the change "behind" a configuration option?  Or perhaps make the change, and provide a configuration option to override the strict 5280 behavior?  The latter offers a rough gauge of how frequently such certs appear in the wild (e.g., discussions or questions about it).

**Reply by tobiasbrunner** [reply]
> Regarding certificates in the wild, would it be safe to put the change "behind" a configuration option? Or perhaps make the change, and provide a configuration option to override the strict 5280 behavior?

We didn't add an option for the similar RFC 4945 compliance change (enforces that end-entity certificates have either no keyUsage extension or that it has at least one of the digitalSignature or nonRepuditation bits set) and that didn't cause issues. So I'd wait with adding one until it's necessary. And as you said, it's reasonable to assume that non-complying CAs have expired by now (or were otherwise replaced e.g. due to the SHA-1 phase-out).

**Reply by tobiasbrunner** [reply]
The changes are now in master.

**Reply by edmorris-github** [reply]
Thank you very much, Tobias.  This will be a great help to many!
